src/coreclr/jit/llvm.cpp

```diff
@@ -759,62 +806,156 @@ Value* buildCall(llvm::IRBuilder<>& builder, GenTree* node)
-    return mapGenTreeToValue(node,
-                             castIfNecessary(builder, builder.CreateLoad(
+    mapGenTreeToValue(node, castIfNecessary(builder, builder.CreateLoad(
                                  castIfNecessary(builder, ptr,
-                                                 getLLVMTypeForVarType(node->TypeGet())->getPointerTo())), getLLVMTypeForVarType(genActualType(node->TypeGet()))));
+                                     getLLVMTypeForVarType(node->TypeGet())->getPointerTo())), getLLVMTypeForVarType(genActualType(node->TypeGet()))));
+}
+
+Value* buildJTrue(llvm::IRBuilder<>& builder, GenTree* node, Value* opValue)
+{
+    return builder.CreateCondBr(opValue, getLLVMBasicBlockForBlock(_currentBlock->bbJumpDest), getLLVMBasicBlockForBlock(_currentBlock->bbNext));
+}
+
+void buildCmp(llvm::IRBuilder<>& builder, genTreeOps op, GenTree* node, Value* op1, Value* op2)
+{
+    llvm::CmpInst::Predicate llvmPredicate;
+
+    switch (op)
+    {
+        case GT_EQ:
+            llvmPredicate = llvm::CmpInst::Predicate::ICMP_EQ;
+            break;
+        case GT_NE:
+            llvmPredicate = llvm::CmpInst::Predicate::ICMP_NE;
+            break;
+        default:
+            failFunctionCompilation(); // TODO all genTreeOps values
+
+    }
+    // comparing refs and ints is valid LIR, but not LLVM so handle that case by converting the int to a ref
+    if (op1->getType() != op2->getType())
+    {
+        if (op1->getType()->isPointerTy() && op2->getType()->isIntegerTy())
+        {
+            op2 = builder.CreateIntToPtr(op2, op1->getType());
+        }
+        else if (op2->getType()->isPointerTy() && op1->getType()->isIntegerTy())
+        {
+            op1 = builder.CreateIntToPtr(op1, op2->getType());
+        }
+        else
+        {
+            // TODO: other valie LIR comparisons
+            failFunctionCompilation();
+        }
+    }
+    mapGenTreeToValue(node, builder.CreateCmp(llvmPredicate, op1, op2));
 }
 
-Value* buildNe(llvm::IRBuilder<>& builder, GenTree* node, Value* op1, Value* op2)
+void buildPhi(llvm::IRBuilder<>& builder, GenTreePhi* phi)
 {
-    return mapGenTreeToValue(node, builder.CreateICmpNE(op1, op2));
+    llvm::PHINode* llvmPhiNode = nullptr;
+    unsigned i = 0;
+    bool requiresLoad = false;
+    for (GenTreePhi::Use& use : phi->Uses())
+    {
+        assert(use.GetNode()->OperIs(GT_PHI_ARG));
+        GenTreePhiArg* phiArg = use.GetNode()->AsPhiArg();
+        unsigned       lclNum = phiArg->GetLclNum();
+        unsigned       ssaNum = phiArg->GetSsaNum();
+
+        LocatedLlvmValue localPhiArg = getSsaLocalForPhi(lclNum, ssaNum);
+        if (i == 0)
+        {
+            requiresLoad = localPhiArg.valueRequiresLoad();
+            Type* phiType = getLLVMTypeForVarType(phi->TypeGet());
+
+            llvmPhiNode = builder.CreatePHI(requiresLoad ? phiType->getPointerTo() : phiType, phi->NumChildren());
```

Comment:
`NumChildren` does linear counting, making the loop quadratic. It should be cached instead.